libs/utils/energy_model.py

```diff
@@ -773,6 +773,216 @@ def _find_core_groups(cls, target):
+                     'This interface is temporary and may change in future '
+                     'according to upstream feedback.')
+
+        if 'cpufreq' not in target.modules:
+            raise TargetError('Requires cpufreq devlib module. Please ensure '
+                               '"cpufreq" is listed in your target/test modules')
+        if 'cpuidle' not in target.modules:
+            raise TargetError('Requires cpuidle devlib module. Please ensure '
+                               '"cpuidle" is listed in your target/test modules')
+
+        # Simplified EM on-disk format (for each frequency domain):
+        #    /sys/devices/system/cpu/energy_model/<frequency_domain>/..
+        #        ../capacity
+        #           contains a space-separated list of capacities in increasing order
+        #        ../cpus
+        #           cpulist-formatted representation of the cpus in the frequency domain
+        #        ../frequency
+        #           space-separated list of frequencies in corresponding order to capacities
+        #        ../power
+        #           space-separated list of power consumption in corresponding order to capacities
+        # taken together, the contents of capacity, frequency and power give you the required
+        # tuple for ActiveStates.
+        # hence, domain should be supplied as a glob, and fields should be
+        #   capacity, cpus, frequency, power
+
+        def sge_path(freq_domain, field):
+            f = '/sys/devices/system/cpu/energy_model/fd{}/{}'
+            return f.format(freq_domain, field)
+
+        # Read all the files we might need in one go, otherwise this will take
+        # ages.
+        sge_globs = [sge_path('**', 'capacity'),
+                     sge_path('**', 'cpus'),
+                     sge_path('**', 'frequency'),
+                     sge_path('**', 'power')]
+        sge_file_values = read_multiple_oneline_files(target, sge_globs)
+
+        if not sge_file_values:
+            raise TargetError('Energy Model not exposed in sysfs. '
+                              'Check CONFIG_SCHED_DEBUG is enabled.')
+        #for key in sge_file_values:
+        #    _log.info('file {}: {}'.format(key, sge_file_values[key]))
+
+        domain_match = re.compile('/sys/devices/system/cpu/energy_model/fd(\d+)/.*')
+        available_sge_freq_domains=[]
+        try:
+            domains = []
+            for key in sge_file_values:
+                m = domain_match.match(key)
+                if m:
+                    domain = m.group(1)
+                    if domain not in domains:
+                        domains.append(domain)
+            available_sge_freq_domains=domains
+        except:
+            pass
+        #_log.info("Available domains: {}".format(available_sge_freq_domains))
+
+        # convert a comma-separated list with ranges to a plain list
+        # e.g. '0,3-5' -> [ '0', '3', '4', '5' ]
+        def cpu_csl_to_list(cpu_csl):
```

Comment:
AFAICT that's already implemented by **devlib.utils.misc.ranges_to_list** (there's a lot of useful "str thing to cpu list" functions in there, when in doubt it's good to have a look)

Reply:
I spent at least half an hour searching for various combinations of list and cpu but missed it. I'll switch over.